processor/src/debug.rs

```diff
@@ -0,0 +1,61 @@
+use crate::{ExecutionError, Felt, Process};
+use core::ops::RangeInclusive;
+use vm_core::Word;
+
+/// VmState holds a current process state information at a specific clock cycle.
+#[derive(Clone, Debug)]
+pub struct VmState {
+    pub clk: usize,
+    pub fmp: Felt,
+    pub stack: Vec<Felt>,
+    pub memory: Vec<(u64, Word)>,
+}
+
+/// Iterator that iterates through vm state at each step of the execution.
+/// This allows debuging or replaying ability to view various process state
+/// at each clock cycle.
+/// If the execution returned an error, it returns that error on the clock cycle
+/// it stopped.
+pub struct VmStateIterator {
+    process: Process,
+    error: Option<ExecutionError>,
+    current_clk: usize,
+}
+
+impl VmStateIterator {
+    pub(super) fn new(process: Process, result: Result<(), ExecutionError>) -> Self {
+        Self {
+            process,
+            error: result.err(),
+            current_clk: 0,
+        }
+    }
+}
+
+impl Iterator for VmStateIterator {
+    type Item = Result<VmState, ExecutionError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_clk > self.process.system.clk() {
+            let error = core::mem::take(&mut self.error);
+            match error {
```

Comment:
Could we say "match self.error" here? Or would it cause some issues?

Reply:
I tried this before, and this is the error I get ->

error[E0507]: cannot move out of `self.error.0` which is behind a mutable reference
  --> processor/src/debug.rs:41:19
   |
41 |             match self.error {
   |                   ^^^^^^^^^^ help: consider borrowing here: `&self.error`
42 |                 Some(e) => return Some(Err(e)),
   |                      -
   |                      |
   |                      data moved here
   |                      move occurs because `e` has type `ExecutionError`, which does not implement the `Copy` trait


If I do match &self.error instead ->

error[E0507]: cannot move out of `self.error.0` which is behind a mutable reference
  --> processor/src/debug.rs:41:19
   |
41 |             match self.error {
   |                   ^^^^^^^^^^ help: consider borrowing here: `&self.error`
42 |                 Some(e) => return Some(Err(e)),
   |                      -
   |                      |
   |                      data moved here
   |                      move occurs because `e` has type `ExecutionError`, which does not implement the `Copy` trait

Reply:
Actually I fixed this by just doing clone instead. Let me know what you think

Reply:
fixed